libs/@local/hashql/core/src/span/table.rs

```diff
@@ -0,0 +1,613 @@
+    ///
+    /// // Table is ready for span insertion
+    /// ```
+    #[must_use]
+    pub const fn new(source: SourceId) -> Self {
+        assert!(
+            source.value() <= SpanId::MAX_SOURCE_ID,
+            "source ID exceeds maximum"
+        );
+
+        Self {
+            source_id: source,
+            spans: Vec::new(),
+            ancestors: Vec::new(),
+        }
+    }
+
+    /// Inserts a new span into the table with the specified ancestors.
+    ///
+    /// Creates a new [`SpanId`] that references the inserted span data and stores the ancestor
+    /// relationships for later resolution. The returned [`SpanId`] encodes both the table's
+    /// source ID and the insertion index.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the there are more spans than which can be represented by the [`SpanId`] type.
+    ///
+    /// # Examples
+    ///
+    /// Inserting a root span:
+    ///
+    /// ```rust
+    /// use hashql_core::span::{SpanAncestors, SpanTable, TextRange};
+    /// use hashql_diagnostics::source::SourceId;
+    ///
+    /// # struct MySpan { range: TextRange }
+    /// # impl hashql_core::span::Span for MySpan {
+    /// #     fn range(&self) -> TextRange { self.range }
+    /// # }
+    /// let mut table = SpanTable::new(SourceId::new_unchecked(0));
+    ///
+    /// let span = MySpan {
+    ///     range: TextRange::new(0.into(), 10.into()),
+    /// };
+    /// let span_id = table.insert(span, SpanAncestors::empty());
+    ///
+    /// // Verify span was inserted successfully
+    /// assert_eq!(
+    ///     table.get(span_id).map(|span| span.range),
+    ///     Some(TextRange::new(0.into(), 10.into()))
+    /// );
+    /// ```
+    ///
+    /// Inserting a span with ancestors:
+    ///
+    /// ```rust
+    /// use hashql_core::span::{SpanAncestors, SpanTable, TextRange};
+    /// use hashql_diagnostics::source::SourceId;
+    ///
+    /// # struct MySpan { range: TextRange }
+    /// # impl hashql_core::span::Span for MySpan {
+    /// #     fn range(&self) -> TextRange { self.range }
+    /// # }
+    /// let mut table = SpanTable::new(SourceId::new_unchecked(1));
+    ///
+    /// // Insert parent span
+    /// let parent = MySpan {
+    ///     range: TextRange::new(10.into(), 50.into()),
+    /// };
+    /// let parent_id = table.insert(parent, SpanAncestors::empty());
+    ///
+    /// // Insert child span with parent as ancestor
+    /// let child = MySpan {
+    ///     range: TextRange::new(5.into(), 15.into()),
+    /// };
+    /// let child_id = table.insert(child, SpanAncestors::union(&[parent_id]));
+    ///
+    /// // Verify child span was inserted successfully
+    /// assert_eq!(
+    ///     table.get(child_id).map(|span| span.range),
+    ///     Some(TextRange::new(5.into(), 15.into()))
+    /// );
+    /// ```
+    #[expect(
+        clippy::cast_possible_truncation,
+        reason = "The arena is not expected to be larger than u32::MAX + debug assertions"
```

Comment:
The assertion is not a debug assertion. Also, the assertion does not help in this context. If `MAX_ID` is `u32::MAX` and `index` is greater than `u32::MAX` the cast will wrap and the assertion does not throw.